src/pubsub.js

```diff
@@ -50,6 +62,94 @@ class BasicPubSub extends Pubsub {
     })
   }
 
+  /**
+   * Dial a received peer.
+   * @override
+   * @param {PeerInfo} peerInfo The peer being dialed
+   * @param {function} callback
+   *
+   * @returns {void}
+   */
+  _dialPeer (peerInfo, callback) {
+    const onDial = (idB58Str, pubsubStopped, conn) => {
+      // pubsub has been stopped, so we should just bail out
+      if (pubsubStopped) {
+        this.log('pubsub was stopped, not processing dial to %s', idB58Str)
+        return callback()
+      }
+
+      this._onDial(peerInfo, conn, callback)
+    }
+
+    callback = callback || function noop () { }
+    const idB58Str = peerInfo.id.toB58String()
+
+    // If already have a PubSub conn, ignore
+    const peer = this.peers.get(idB58Str)
+    if (peer && peer.isConnected) {
+      return nextTick(() => callback())
+    }
+
+    // If already dialing this peer, ignore
+    if (this._dials.has(idB58Str)) {
+      this.log('already dialing %s, ignoring dial attempt', idB58Str)
+      return nextTick(() => callback())
+    }
+
+    // Verify if is known that the peer does not support Gossipsub
+    let notSupportingGossipsub = peerInfo.protocols.size && !peerInfo.protocols.has(this.multicodec)
```

Comment:
I think this might be a little clearer. If it has the floodsub protocol and not gossipsub, we know it doesn't support the latter. I changed the variable name because the negative name was tripping me up reading the code.

```suggestion
    let onlySupportsFloodsub = peerInfo.protocols.has(floodsubMulticodec) && !peerInfo.protocols.has(this.multicodec)
```